verifiable-db/src/revelation/revelation_unproven_offset.rs

```diff
@@ -0,0 +1,1295 @@
+                let row_tree_root = row_path_wires.root;
+                // compute hash of the index node storing the rows tree containing the current row
+                let index_node_hash = {
+                    let inputs = index_node_info[i]
+                        .child_hashes
+                        .into_iter()
+                        .flat_map(|hash| hash.to_targets())
+                        .chain(index_node_info[i].node_min.to_targets())
+                        .chain(index_node_info[i].node_max.to_targets())
+                        .chain(once(index_ids[0]))
+                        .chain(row_proof.index_value_target().to_targets())
+                        .chain(row_tree_root.to_targets())
+                        .collect_vec();
+                    b.hash_n_to_hash_no_pad::<H>(inputs)
+                };
+                let index_path_wires = MerklePathGadget::build(b, index_node_hash, index_ids[0]);
+                // if the current row is valid, check that the root is the same of the original tree, completing
+                // membership proof for the current row; otherwise, we don't care
+                let root = b.select_hash(is_matching_row, &index_path_wires.root, &tree_hash);
+                b.connect_hashes(tree_hash, root);
+
+                row_paths.push(row_path_wires.inputs);
+                index_paths.push(index_path_wires.inputs);
+                // check that the primary index value for the current row is within the query
+                // bounds (only if the row is valid)
+                let index_value = row_proof.index_value_target();
+                let greater_than_min = b.is_less_or_equal_than_u256(&min_query, &index_value);
+                let smaller_than_max = b.is_less_or_equal_than_u256(&index_value, &max_query);
+                let in_range = b.and(greater_than_min, smaller_than_max);
+                let in_range = b.and(is_matching_row, in_range);
+                b.connect(in_range.target, is_matching_row.target);
+
+                // enforce DISTINCT only for actual results: we enforce the i-th actual result is strictly smaller
+                // than the (i+1)-th actual result
+                max_result = if let Some(res) = &max_result {
+                    let current_result: [UInt256Target; S] =
+                        get_result(i).to_vec().try_into().unwrap();
+                    let is_smaller = b.is_less_than_or_equal_to_u256_arr(res, &current_result).0;
+                    // flag specifying whether we must enforce DISTINCT for the current result or not
+                    let must_be_enforced = b.and(is_matching_row, distinct);
+                    let is_smaller = b.and(must_be_enforced, is_smaller);
+                    b.connect(is_smaller.target, must_be_enforced.target);
+                    Some(current_result)
+                } else {
+                    Some(get_result(i).to_vec().try_into().unwrap())
+                };
```

Comment:
have you considered using the digest of the results and just making sure they're all different ? Since checking the digest would only entail 1 digest selection accross the S rows, instead of checking all the results for each row.
Might be worse or better depending on the constants used i guess ?

Reply:
Yes, I considered it but I didn't do it eventually because I would have needed to add a comparison gadget for digests, while for vector of `u256` it was already available. So, given that we had a strict timeline, I decided to cut the scope and do like this. But should be actually more efficient to implement this logic over digests rather than all the results, so if you think it's ok to spend some time doing it (i.e., should take like half a day at most), I can definitely do it. Wdyt?

Reply:
Actually I realized that to have a comparison for digests, we would need to split each of the 11 field elements of the digest in 32 bit limbs, to employ the same order relationship we currently employ for slices of 32 bit limbs. This split would be done with [this method](https://github.com/Lagrange-Labs/mapreduce-plonky2/blob/feat/unproven-limit-offset-queries/mp2-common/src/poseidon.rs#L56), which however requires about 2 rows per field element (due to range-checks). So, to split a digest into 32 bit limbs, we would need 22 Plonky2 rows. Then, computing the comparison with another digest (already split in 32 bit limbs) will require other 8 rows, so 30 rows in total.
On the other hand, with the current comparison of the results, we need about 3 rows per each result item, so with the same number of rows we would be able to compare up to 10 result items, which looks more than enough to me. 
So, TLDR, comparing digests instead of the actual results would become convenient only if we have more than 10 result items per row (i.e., more than 10 items specified in the `SELECT` statement of the query), so it doesn't seem really worthy to me, given that it will also require additional work to implement this comparison. Wdyt?

Reply:
Then it's problem solved :) THanks !